src/config/mod.rs

```diff
@@ -3995,21 +3981,27 @@ impl TikvConfig {
             self.rocksdb.raftcf.disable_write_stall = true;
         }
         // Fill in values for unspecified write stall configurations.
+
         macro_rules! fill_cf_opts {
             ($cf_opts:expr, $cfg:expr) => {
-                if let Some(v) = &mut $cf_opts.level0_slowdown_writes_trigger {
-                    if $cfg.enable && *v > $cfg.l0_files_threshold as i32 {
-                        warn!(
-                            "{}.level0-slowdown-writes-trigger is too large. Setting it to \
-                            storage.flow-control.l0-files-threshold ({})",
-                            stringify!($cf_opts), $cfg.l0_files_threshold
-                        );
-                        *v = $cfg.l0_files_threshold as i32;
-                    }
-                } else {
-                    $cf_opts.level0_slowdown_writes_trigger =
-                        Some($cfg.l0_files_threshold as i32);
+                // Automatically override `level0_slowdown_writes_trigger` if it is set too high,
+                // to ensure that RocksDB's compaction speed-up mechanism can help avoid triggering flow control,
+                // since RocksDB starts speed up compaction when the number of L0 files reaches this threshold.
+                //
+                // For the same reason, when `l0-file-threshold` is set to a higher value, we do **not** override it,
+                // because we still want RocksDB's compaction speed-up mechanism to take effect—
+                // allowing compaction to proceed more quickly and smoothly, leading to more stable performance
+                // and avoiding issues like https://github.com/tikv/tikv/issues/18708.
+                if $cfg.enable && $cf_opts.level0_slowdown_writes_trigger > $cfg.l0_files_threshold as i32 {
+                    warn!(
+                        "{}.level0-slowdown-writes-trigger is too large. Setting it to \
+                        storage.flow-control.l0-files-threshold ({})",
+                        stringify!($cf_opts), $cfg.l0_files_threshold
+                    );
+                    $cf_opts.level0_slowdown_writes_trigger = $cfg.l0_files_threshold as i32;
                 }
+                // Keep override for level0_stop_writes_trigger to ensure ingest_maybe_stall
```

Comment:
nit: this may highlight their differences?
```suggestion
                // If unset, `level0_stop_writes_trigger` defaults to `l0_files_threshold` (unlike `level0_slowdown_writes_trigger`, which defaults to 20) to ensure ingest_maybe_stall
```

Reply:
done